### PR TITLE
Added firewall app

### PR DESCRIPTION
**Pull request by Mixaster995**
Signed-off-by: Mikhail Avramenko <avramenkomihail15@gmail.com>

## Description
Made firewall nse, which filters traffic according to acl rules, provided in config file
Depends of https://github.com/networkservicemesh/sdk-vpp/pull/291 

## Issue link
fixes https://github.com/networkservicemesh/sdk/issues/952

## How Has This Been Tested?
- [ ] Added unit testing to cover
- [x] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested
## Types of changes
- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -14,7 +14,312 @@
...
+	log.FromContext(ctx).Infof("1: get config from environment")
+	log.FromContext(ctx).Infof("2: retrieve spiffe svid")
+	log.FromContext(ctx).Infof("3: create grpc client options")
+	log.FromContext(ctx).Infof("4: create firewall network service endpoint")
+	log.FromContext(ctx).Infof("5: create grpc and mount nse")
+	log.FromContext(ctx).Infof("6: register nse with nsm")
+	log.FromContext(ctx).Infof("a final success message with start time duration")
+}
+
+func setupLogging(ctx context.Context) {
```
Comment (denis-tingaikin) [inline main.go:242]: this function is incorrect...
We should use context with logger in main function.

Please take a look at what is structure data context: https://golang.org/pkg/context/#WithValue

File: main.go [hunk 2]
```diff
@@ -14,7 +14,312 @@
...
+	if err := config.Process(); err != nil {
+		logrus.Fatal(err.Error())
+	}
+
+	log.FromContext(ctx).Infof("Config: %#v", config)
+
+	return config
+}
+
+func retrieveSvid(ctx context.Context) *workloadapi.X509Source {
```
Comment (denis-tingaikin) [inline main.go:269]: Please use this inline.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]
Other things look good.

This pr is blocked by https://github.com/networkservicemesh/sdk-vpp/pull/291

File: main.go (hunk 1 above)
Comment (Mixaster995) [reply main.go]: ok, fixed

File: go.mod [hunk 3]
```diff
@@ -1,3 +1,21 @@
...
+	github.com/edwarnicke/grpcfd v0.1.0
+	github.com/edwarnicke/vpphelper v0.0.0-20210512223648-f914b171f679
+	github.com/kelseyhightower/envconfig v1.4.0
+	github.com/networkservicemesh/api v1.0.0
+	github.com/networkservicemesh/sdk v1.0.0
+	github.com/networkservicemesh/sdk-vpp v0.0.0-20210630112222-f9cd27fdb5e8
+	github.com/pkg/errors v0.9.1
+	github.com/sirupsen/logrus v1.8.1
+	github.com/spiffe/go-spiffe/v2 v2.0.0-beta.6
+	google.golang.org/genproto v0.0.0-20210602131652-f16073e35f0c // indirect
```
[the remark this answers is not in the document]
Comment (Mixaster995) [reply go.mod]: done

File: go.mod [hunk 4]
```diff
@@ -1,3 +1,21 @@
...
+	google.golang.org/grpc v1.38.0
+	gopkg.in/yaml.v2 v2.4.0
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
```
[the remark this answers is not in the document]
Comment (Mixaster995) [reply go.mod]: done

File: main.go (hunk 2 above)
Comment (Mixaster995) [reply main.go]: done

File: main.go [hunk 5]
```diff
@@ -14,7 +14,312 @@
...
+	svid, err := source.GetX509SVID()
+	if err != nil {
+		logrus.Fatalf("error getting x509 svid: %+v", err)
+	}
+	log.FromContext(ctx).Infof("SVID: %q", svid.ID)
+
+	return source
+}
+
+func getEndpoint(ctx context.Context, config *Config, source x509svid.Source, vppConn vpphelper.Connection, clientOptions ...grpc.DialOption) endpoint.Endpoint {
```
[the remark this answers is not in the document]
Comment (Mixaster995) [reply main.go]: done

File: internal/config/config.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+type aclConfig struct {
+	ACLRules map[string]acl_types.ACLRule
+}
+
+// GetACLRules reads config file with rules for acl filtering and return it as map
+func GetACLRules(ctx context.Context) (rules []acl_types.ACLRule) {
+	var resultRules []acl_types.ACLRule
+	logger := log.FromContext(ctx).WithField("acl", "config")
+
+	raw, err := ioutil.ReadFile("/etc/vppagent-firewall/config.yaml")
```
Comment (denis-tingaikin) [inline internal/config/config.go:39]: Please move path to all config into envconfig. Also consider to include alc config into envconfig.
```suggestion
	raw, err := ioutil.ReadFile("/etc/vppagent-firewall/config.yaml")
```

File: main.go [hunk 7]
```diff
@@ -14,7 +14,312 @@
...
+	ServiceName      string            `default:"" desc:"Name of providing service" split_words:"true"`
+	Labels           map[string]string `default:"" desc:"Endpoint labels"`
+}
+
+// Process prints and processes env to config
+func (c *Config) Process() error {
+	if err := envconfig.Usage("nse", c); err != nil {
+		return errors.Wrap(err, "cannot show usage of envconfig nse")
+	}
+	if err := envconfig.Process("nse", c); err != nil {
```
[the remark this answers is not in the document]
Comment (Mixaster995) [reply main.go]: done

File: main.go [hunk 8]
```diff
@@ -14,7 +14,312 @@
...
+	ListenOn         string            `default:"listen.on.sock" desc:"listen on socket" split_words:"true"`
+	ConnectTo        url.URL           `default:"unix:///var/lib/networkservicemesh/nsm.io.sock" desc:"url to connect to" split_words:"true"`
+	MaxTokenLifetime time.Duration     `default:"24h" desc:"maximum lifetime of tokens" split_words:"true"`
+	ServiceName      string            `default:"" desc:"Name of providing service" split_words:"true"`
+	Labels           map[string]string `default:"" desc:"Endpoint labels"`
+}
+
+// Process prints and processes env to config
+func (c *Config) Process() error {
+	if err := envconfig.Usage("nse", c); err != nil {
```
[the remark this answers is not in the document]
Comment (Mixaster995) [reply main.go]: done

File: internal/config/config.go (hunk 6 above)
Comment (Mixaster995) [reply internal/config/config.go]: done

**Review by d-uzlov — APPROVED** [review]

File: internal/config/config.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	"path/filepath"
+
+	"github.com/edwarnicke/govpp/binapi/acl_types"
+	"gopkg.in/yaml.v2"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+type aclConfig struct {
+	ACLRules map[string]acl_types.ACLRule
```
Comment (denis-tingaikin) [inline internal/config/config.go:32]: Can we use this map inline in `Config` ?

File: internal/config/config.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// GetACLRules reads config file with rules for acl filtering and return it as map
+func GetACLRules(ctx context.Context, path string) (rules []acl_types.ACLRule) {
```
Comment (denis-tingaikin) [inline internal/config/config.go:36]: Consider make it a function for `Config`
```suggestion
func (c*Config) RetrieveACLRules(ctx context.Context) (rules []acl_types.ACLRule) {
```

File: main.go [hunk 11]
```diff
@@ -14,7 +14,291 @@
...
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("startup completed in %v", time.Since(starttime))
+	// ********************************************************************************
+
+	// wait for server to exit
+	<-ctx.Done()
+	<-vppErrCh
+}
+
+func exitOnErr(ctx context.Context, cancel context.CancelFunc, errCh <-chan error) {
```
Comment (denis-tingaikin) [inline main.go:280]: @edwarnicke I see this function in many places. Can we move it into sdk/tools?

This can be done in separate PRs.

**Review by denis-tingaikin — COMMENTED** [review]
Added few comments.

In general, looks good.

Also please subscribe this repository for updates in sdk/vpp
https://github.com/networkservicemesh/sdk-vpp/blob/main/.github/workflows/update-dependent-repositories-gomod.yaml#L18

File: internal/config/config.go (hunk 9 above)
Comment (Mixaster995) [reply internal/config/config.go]: ok, done

File: internal/config/config.go (hunk 10 above)
Comment (Mixaster995) [reply internal/config/config.go]: ok, done

**Comment by Mixaster995** [comment]
> Added few comments.
> 
> In general, looks good.
> 
> Also please subscribe this repository for updates in sdk/vpp
> https://github.com/networkservicemesh/sdk-vpp/blob/main/.github/workflows/update-dependent-repositories-gomod.yaml#L18

**Review by denis-tingaikin — APPROVED** [review]

**Review by denis-tingaikin — COMMENTED** [review]
One thing to complete this.

Please add docker caching as in other repos:
https://github.com/networkservicemesh/cmd-nse-supplier-k8s/blob/main/internal/pkg/imports/gen.go#L20-L22

**Comment by denis-tingaikin** [comment]
Please add using of the gocahe. See at example https://github.com/networkservicemesh/cmd-forwarder-vpp/blob/main/Dockerfile#L20-L21

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**